### DISCUSSION TITLE
Empty string in query parameter is considered missing

### DISCUSSION THREAD

**Discussion by VictorBac** [fastapi/fastapi#6850] [Questions] [answered]
Hi team. 

I have an Update endpoint on some data in my backend. 
Every field that is sent to the endpoint will replace the data in DB. So if I send only 3 fields out of 10, only the 3 fields will be updated

But if I send an empty string, fastapi replace it by None.
=> _I can't differentiate a missing parameter and an empty string_.
An empty string means I want to clear this specific field

I tried
```Python
notes: _t.Optional[str] = Form(None),
```

and 
```Python
notes: _t.Optional[str] = None,
```

But every time it is transformed to None. Is there a way to accept and keep empty strings ?
PS: I'm sending data using axios, but I don't think it removes empty string (from what I see from my chrome console)

I am using FastAPI version 0.54.2.

Thank you!

**Comment by krishnardt** [comment]
Above details are not sufficient to define the issue(follow issue template).
Can you share example code?
If you can or your project supports, update fastapi to latest version.

**Comment by VictorBac** [comment]
Sorry I should have given more details.

On latest version of fastapi (0.63.0):

```Python
from typing import Optional

from fastapi import FastAPI, Form

app = FastAPI()

@app.post("/items")
async def create_item(
        item_description: Optional[str] = Form(None),
):
    print(item_description)
    print(item_description is None)
    print(item_description == '')
```

If I post on the endpoint:
```
curl -X POST "http://127.0.0.1:8000/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "item_description="
```

as a result I get:
```
None
True
False
```

I expect item_description to be an empty string, not None. 

I want to be able to differentiate those 2 calls.

No data:
```
curl -X POST "http://127.0.0.1:8001/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d ""
```

Data empty:
```
curl -X POST "http://127.0.0.1:8001/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "item_description="

```

**Comment by krishnardt** [comment]
Have you tried this one?

`curl -X POST "http://127.0.0.1:8001/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "item_description='' "
`
There is a difference between without passing any params value or passing blank param values.



Here..
`curl -X POST "http://127.0.0.1:8001/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d ""
`
you are sending no params... the form will take it as empty and returns None


`curl -X POST "http://127.0.0.1:8001/items" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "item_description="
`
you are sending param... but no values assigned to it, even '' also and the form will take the param value as None.

**Comment by falkben** [comment, marked as answer]
It sounds like you just want the default value to be an empty string?

If that's the case, just change the value that `item_description` is assigned to be Form(""). The first argument of the param function is the default value assigned.

```python
from fastapi import FastAPI, Form, testclient

app = FastAPI()


@app.post("/items")
async def create_item(
    item_description: str = Form(""),
):
    print(item_description)
    print(item_description is None)
    print(item_description == "")
```

**Comment by krishnardt** [comment]
yes, it depends on the default value.

**Comment by VictorBac** [comment]
I can't use ```item_description: str = Form(""),``` because it wouldn't differentiate a missing parameter from a parameter with empty string.

> Have you tried this one?
> 
> `curl -X POST "http://127.0.0.1:8001/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "item_description='' " `
> There is a difference between without passing any params value or passing blank param values.
> ....
> `curl -X POST "http://127.0.0.1:8001/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "item_description=" `
> you are sending param... but no values assigned to it, even '' also and the form will take the param value as None.

Ok, it seems to be working ! **I was missing the empty string in my data,** so it was defaulting to None.

There is indead a difference between:
```-d "item_description=''"``` => I get the empty string in my variable

and
```-d "item_description=" ``` => defaulting to None


Now I just need to find how to send an empty string with axios. it seems to replace it with empty data, so the backend default to None..

Thank you so much for your help.

**Comment by VictorBac** [comment]
> Have you tried this one?
> 
> `curl -X POST "http://127.0.0.1:8001/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "item_description='' " `
> There is a difference between without passing any params value or passing blank param values.
> 
> Here..
> `curl -X POST "http://127.0.0.1:8001/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "" `
> you are sending no params... the form will take it as empty and returns None
> 
> `curl -X POST "http://127.0.0.1:8001/items" -H "accept: application/json" -H "Content-Type: application/x-www-form-urlencoded" -d "item_description=" `
> you are sending param... but no values assigned to it, even '' also and the form will take the param value as None.

I'm reopening because I find it is a very strange behavior.

Is there no way to differentiate **no param** ```curl -d "" ``` and **blank param values** ```curl -d "item_description="   ``` ?

``` -d "item_description=''"   ```  is really awkward to handle, I have to do something like this:

```Python
if item_description == "''":
    do_something()
```

**Comment by ycd** [comment]
> -d "item_description=''" is really awkward to handle

It is awkward indeed :rofl: 

Wouldn't it be better if you handle all the validation logic inside the Pydantic model?

**Comment by krishnardt** [comment]
Yes, awkward solution for an awkward curl in the description..😉

As @ycd suggested, you can handle it in pydantic class as per your usecase.

**Comment by VictorBac** [comment]
Ok, I don't have this problem when I am using pydantic objects

something like this:
```Python
from pydantic import BaseModel, validator

class Item(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None

@app.post("/items")
async def create_item(
        item: Item,
):
    return
```

The data send in the request is a json, and I can ommit some field or send empty value.
```curl ..... -d "{\"name\":\"\"}"```

```name``` would be emptied, and ```description``` left untouched ! 

Thank you all for your help!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @VictorBac  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by grubberr** [comment]
I agree with the thread starter @VictorBac

That in FastAPI (`fastapi==0.111.0, pydantic==2.7.1`) in the `Form()` parameter there is no way to distinguish between 2 states:

1) no parameter at all
`curl -d ''`.

2) parameter passed but without value
`curl -d 'key='`.

In both cases I get the value `None` or empty string if I change the default.

**The main problem is that these 2 states are mixed together and I cannot distinguish them.**

For example, my old backend application written on `flask-restx` has such a possibility to differentiate these 2 states.

**Comment by y2kbugger** [comment]
htmx sends unquoted form fields and so when a textbox is empty, we get missing field instead of empty string. greenfield is not so bad, but it seems awkward to have to give a default value for a required field if empty string is a valid value.  I'd like to reserve 422 for when we actually have a bug and the field is actually missing.
